File: frontend/services/api/manualEntries.ts
import { ManualEntry, DateRange } from '../../types';
import { invoke, dateRangeToParams } from './utils';

export const manualEntriesApi = {
  getManualEntries: (range: DateRange): Promise<ManualEntry[]> => {
    return invoke('get_manual_entries', dateRangeToParams(range));
  },
  
  createManualEntry: (entry: {
    description: string | null;
    category_id: number | null;
    started_at: number;
    ended_at: number;
    // Плагины должны использовать call_db_method для работы с этими полями
  }): Promise<ManualEntry> => {
    return invoke('create_manual_entry', {
      description: entry.description,
      categoryId: entry.category_id,
      startedAt: entry.started_at,
      endedAt: entry.ended_at,
    });
  },
  
  updateManualEntry: (entry: {
    id: number;
    description: string | null;
    category_id: number | null;
    started_at: number;
    ended_at: number;
    // Плагины должны использовать call_db_method для работы с этими полями
  }): Promise<ManualEntry> => {
    return invoke('update_manual_entry', {
      id: entry.id,
      description: entry.description,
      categoryId: entry.category_id,
      startedAt: entry.started_at,
      endedAt: entry.ended_at,
    });
  },
  
  deleteManualEntry: (id: number): Promise<void> => {
    return invoke('delete_manual_entry', { id });
  },
  
  startManualEntry: (categoryId: number, description?: string): Promise<number> => {
    return invoke('start_manual_entry', { categoryId, description });
  },
  
  stopManualEntry: (): Promise<ManualEntry> => {
    return invoke('stop_manual_entry');
  },
};

File: backend/src/database/manual_entries.rs
//! Manual entry database operations

use rusqlite::{Result, params};
use super::common::Database;
use super::models::ManualEntry;

impl Database {
    /// Add a manual entry
    pub fn add_manual_entry(
        &self,
        description: Option<&str>,
        category_id: Option<i64>,
        started_at: i64,
        ended_at: i64,
    ) -> Result<i64> {
        let conn = self.conn.lock().unwrap();
        conn.execute(
            "INSERT INTO manual_entries (entry_type, description, category_id, started_at, ended_at)
             VALUES ('', ?, ?, ?, ?)",
            params![description, category_id, started_at, ended_at],
        )?;
        Ok(conn.last_insert_rowid())
    }

    /// Get manual entries for a time range
    pub fn get_manual_entries(&self, start: i64, end: i64) -> Result<Vec<ManualEntry>> {
        let conn = self.conn.lock().unwrap();
        let mut stmt = conn.prepare(
            "SELECT id, description, category_id, started_at, ended_at
             FROM manual_entries
             WHERE started_at >= ? AND started_at <= ?
             ORDER BY started_at ASC",
        )?;

        let entries = stmt
            .query_map(params![start, end], |row| {
                Ok(ManualEntry {
                    id: row.get(0)?,
                    description: row.get(1)?,
                    category_id: row.get(2)?,
                    started_at: row.get(3)?,
                    ended_at: row.get(4)?,
                })
            })?
            .collect::<Result<Vec<_>>>()?;

        Ok(entries)
    }

    /// Update manual entry
    pub fn update_manual_entry(
        &self,
        id: i64,
        description: Option<&str>,
        category_id: Option<i64>,
        started_at: i64,
        ended_at: i64,
    ) -> Result<()> {
        let conn = self.conn.lock().unwrap();
        conn.execute(
            "UPDATE manual_entries SET entry_type = '', description = ?, category_id = ?, 
             started_at = ?, ended_at = ? WHERE id = ?",
            params![description, category_id, started_at, ended_at, id],
        )?;
        Ok(())
    }
    

    /// Delete manual entry
    pub fn delete_manual_entry(&self, id: i64) -> Result<()> {
        let conn = self.conn.lock().unwrap();
        conn.execute("DELETE FROM manual_entries WHERE id = ?", params![id])?;
        Ok(())
    }
}
